File: backend/core/internal/middleware/rate_limiter.go
package middleware

import (
	"fmt"
	"net/http"
	"time"

	"bank-core/internal/logger"
	"github.com/gin-gonic/gin"
	"github.com/redis/go-redis/v9"
)

// RedisSlidingWindowRateLimiter implements a sliding window counter algorithm using Redis Sorted Sets (ZSET).
// This guarantees strict rate limit accuracy across distributed microservice instances.
func RedisSlidingWindowRateLimiter(rdb *redis.Client, maxRequests int, windowDuration time.Duration) gin.HandlerFunc {
	return func(c *gin.Context) {
		if rdb == nil {
			c.Next()
			return
		}

		ctx := c.Request.Context()
		clientIP := c.ClientIP()
		key := fmt.Sprintf("rate_limit:%s:%s", c.FullPath(), clientIP)
		now := time.Now().UnixNano()
		windowStart := now - windowDuration.Nanoseconds()

		pipe := rdb.Pipeline()
		// 1. Remove elements older than the current sliding window
		pipe.ZRemRangeByScore(ctx, key, "-inf", fmt.Sprintf("%d", windowStart))
		// 2. Count current elements in window
		cardCmd := pipe.ZCard(ctx, key)
		// 3. Add current request timestamp
		pipe.ZAdd(ctx, key, redis.Z{Score: float64(now), Member: fmt.Sprintf("%d", now)})
		// 4. Set TTL for the key
		pipe.Expire(ctx, key, windowDuration)

		_, err := pipe.Exec(ctx)
		if err != nil {
			logger.Warn(ctx, "Redis rate limiter error, allowing request by fail-open policy", map[string]interface{}{
				"error": err.Error(),
			})
			c.Next()
			return
		}

		currentCount := cardCmd.Val()
		if currentCount >= int64(maxRequests) {
			logger.Warn(ctx, "Rate limit exceeded (sliding window)", map[string]interface{}{
				"ip":           clientIP,
				"path":         c.Request.URL.Path,
				"limit":        maxRequests,
				"window_sec":   windowDuration.Seconds(),
				"current_reqs": currentCount,
			})

			c.Header("Retry-After", fmt.Sprintf("%d", int(windowDuration.Seconds())))
			c.JSON(http.StatusTooManyRequests, gin.H{
				"error":       "Too many requests. Please slow down.",
				"retry_after": int(windowDuration.Seconds()),
			})
			c.Abort()
			return
		}

		c.Header("X-RateLimit-Limit", fmt.Sprintf("%d", maxRequests))
		c.Header("X-RateLimit-Remaining", fmt.Sprintf("%d", int64(maxRequests)-currentCount-1))
		c.Next()
	}
}

File: backend/ai-service/app/middleware/rate_limiter.py
import time
import logging
from typing import Optional
import redis.asyncio as aioredis
from fastapi import HTTPException, Request

logger = logging.getLogger("ai_service.rate_limiter")

class RedisSlidingWindowRateLimiter:
    def __init__(self, redis_url: str, max_requests: int = 60, window_sec: int = 60):
        self.redis_url = redis_url
        self.max_requests = max_requests
        self.window_sec = window_sec
        self.redis: Optional[aioredis.Redis] = None

    async def connect(self):
        try:
            self.redis = aioredis.from_url(self.redis_url, decode_responses=True)
            await self.redis.ping()
            logger.info("Connected to Redis for sliding window rate limiter")
        except Exception as e:
            logger.warning(f"Redis unavailable, rate limiter running in bypass mode: {e}")
            self.redis = None

    async def check(self, request: Request, identifier: Optional[str] = None):
        if not self.redis:
            return

        client_ip = identifier or request.client.host or "127.0.0.1"
        key = f"rate_limit:ai:{client_ip}"
        now = time.time()
        clear_before = now - self.window_sec

        try:
            pipe = self.redis.pipeline()
            pipe.zremrangebyscore(key, 0, clear_before)
            pipe.zadd(key, {str(now): now})
            pipe.zcard(key)
            pipe.expire(key, self.window_sec + 10)
            results = await pipe.execute()

            current_requests = results[2]
            if current_requests > self.max_requests:
                logger.warn(f"Rate limit exceeded for client {client_ip} ({current_requests}/{self.max_requests})")
                raise HTTPException(
                    status_code=429,
                    detail=f"Too Many Requests. Rate limit of {self.max_requests} req/min exceeded. Please try again in a moment."
                )
        except HTTPException:
            raise
        except Exception as e:
            logger.error(f"Error executing Redis rate limiter: {e}")
